### PR TITLE
Events and permissions

### PR DESCRIPTION
**Pull request by clarkwinkelmann**
My original intent was to add the event that's missing. Core still has https://github.com/flarum/core/blob/d492579638fb52dafbfe65f1f36a95eb6047f7f3/src/User/Event/BioChanged.php but it's not used anywhere. So I created a new event with the same signature inside of this extension.

While doing that I updated the npm config, npm dependencies and ran the linter.

I also noticed the odd mess that we used to handle suspended users. It makes much more sense to go through a policy, that way the suspend extension can automatically handle it.

While I was adding the policy, it was much simpler to just create dedicated permissions for the extension, which now gives a lot more flexibility.

My intent would be to release this as version 2.0 of user-bio.

### CODE REVIEW

File: src/Access/UserPolicy.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+use Flarum\User\User;
+
+class UserPolicy extends AbstractPolicy
+{
+    protected $model = User::class;
+
+    public function viewBio(User $actor, User $user)
+    {
+        // We only let the user see its own bio if they are also allowed to edit it
+        //var_dump($actor->hasPermission('fof-user-bio.editOwn'));die();
```
Comment (imorland) [inline src/Access/UserPolicy.php]: var_dump ;)

File: js/package.json [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
+        "webpack": "^4.0.0",
+        "webpack-cli": "^3.0.7"
+    },
+    "scripts": {
+        "dev": "webpack --mode development --watch",
+        "build": "webpack --mode production",
+        "lint": "prettier --single-quote --trailing-comma es5 --print-width 150 --tab-width 4 --write 'src/**/*'"
+    },
+    "devDependencies": {
+        "prettier": "^1.15.2"
```
Comment (imorland) [inline js/package.json:16]: Would prettier v2 be a more appropriate choice?

**Review by imorland — COMMENTED** [review]
A much needed overhaul!

My only big thing, and it's a question really, should the `bio` content be passed thru the `Formatter`? Could allow for a little extra flexibility. Of course, then a migration would be needed to process the any existing bios, but...

File: src/Access/UserPolicy.php (hunk 1 above)
Comment (clarkwinkelmann) [reply src/Access/UserPolicy.php]: ooops

File: js/package.json (hunk 2 above)
Comment (clarkwinkelmann) [reply js/package.json:16]: Not sure. I've kept the original configuration that was already in the package (but it hadn't be used it seems so...)

**Comment by clarkwinkelmann** [comment]
@imorland yep I also think it would be nice to use the formatter at some point. But I'd say we keep that for a next update. This update doesn't require any migration.

The only breaking-but-not-really-change is that the new permissions won't automatically reflect the default from 0.1. But that's why I will be bumping to 0.2 and mention that in the release notes. I don't think it's worth adding a migration for that. Also, people might want to take the opportunity to change the "view" permission anyway.

File: js/src/forum/components/UserBio.js [hunk 3]
```diff
@@ -8,97 +8,113 @@ import extractText from 'flarum/utils/extractText';
...
+        } else {
+            let subContent;
+
+            if (this.loading) {
+                subContent = <p className="UserBio-placeholder">{LoadingIndicator.component({ size: 'tiny' })}</p>;
+            } else {
+                const bioHtml = user.bioHtml();
+
+                if (bioHtml) {
+                    subContent = m.trust(bioHtml);
```
Comment (imorland) [inline js/src/forum/components/UserBio.js:50]: Would using `m.trust` here open the _potential_ to place malicious code/any HTML in the bio content, especially without the `Formatter`?

Not sure, but thought it worth raising..

**Comment by imorland** [comment]
@clarkwinkelmann agreed. Not tested locally, but looks good, other than the latest point above

File: js/package.json (hunk 2 above)
Comment (imorland) [reply js/package.json:16]: I guess it does not really matter either way :)

File: js/src/forum/components/UserBio.js (hunk 3 above)
Comment (clarkwinkelmann) [reply js/src/forum/components/UserBio.js:50]: I think it's safe. This is the code used to create `bioHtml()`.

```
User.prototype.bioHtml = computed('bio', bio =>
        bio
            ? '<p>' +
              $('<div/>')
                  .text(bio)
                  .html()
                  .replace(/\n/g, '<br>')
                  .autoLink({ rel: 'nofollow ugc' }) +
              '</p>'
            : ''
    );
```

To my knowledge `.text(text).html()` is safe against HTML injection.

I don't see any risk with the `replace` call either.

Dangerous point might be that `autoLink` library I have not reviewed.

I know some apps use a `template` element to convert text to html. But I'm not sure whether it has any benefit over this.

Comment (imorland) [reply js/src/forum/components/UserBio.js:50]: Ok, great :)

**Review by imorland — APPROVED** [review]

File: js/package.json (hunk 2 above)
Comment (dsevillamartin) [reply js/package.json:16]: It's not a focus of this PR, so it's fine not updating. It'd also mean all files with arrow functions & one arg would have changes, so let's avoid that :)

**Pull request merged**